### PR TITLE
Feat/remove opportunity

### PR DESCRIPTION
**Pull request by danimhr**
This PR aims to add remove opportunity functionality to auction server

### CODE REVIEW

**Comment by m30m** [comment]
Are you going to make the necessary changes in the limonade too?

File: auction-server/src/opportunity/api.rs [hunk 1]
```diff
@@ -722,10 +765,43 @@ pub async fn post_quote(
...
+    Ok(StatusCode::NO_CONTENT)
+}
+
+pub fn get_routes(store: Arc<StoreNew>) -> Router<Arc<StoreNew>> {
     Router::new()
         .route("/", post(post_opportunity))
         .route("/quote", post(post_quote))
         .route("/", get(get_opportunities))
         .route("/:opportunity_id/bids", post(opportunity_bid))
+        .route("/", login_required!(store, delete(delete_opportunities)))
```
[the remark this answers is not in the document]
Comment (danimhr) [reply auction-server/src/opportunity/api.rs:806]: So in this case, we need to run the verification to make sure the opportunity is not valid before removing it

File: auction-server/src/opportunity/repository/remove_opportunities.rs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        &self,
+        db: &sqlx::Pool<Postgres>,
+        permission_key: PermissionKey,
+        reason: OpportunityRemovalReason,
+    ) -> Vec<T::Opportunity> {
+        let mut write_guard = self.in_memory_store.opportunities.write().await;
+        let opportunitties = write_guard.remove(&permission_key);
+        drop(write_guard);
+
+        tokio::spawn({
```
[the remark this answers is not in the document]
Comment (danimhr) [reply auction-server/src/opportunity/repository/remove_opportunities.rs:26]: Will update it to be like remove_opportunity

File: auction-server/src/opportunity/repository/get_live_opportunities.rs [hunk 3]
```diff
@@ -3,19 +3,19 @@ use {
         InMemoryStore,
         Repository,
     },
-    crate::kernel::entities::PermissionKey,
+    crate::opportunity::entities,
 };
 
 impl<T: InMemoryStore> Repository<T> {
-    pub async fn get_live_opportunities_by_permission_key(
+    pub async fn get_live_opportunities(
```
Comment (m30m) [inline auction-server/src/opportunity/repository/get_live_opportunities.rs:10]: can we refactor the name of these functions. we have `get_opportunities`, `get_all_opportunities`, and `get_live_opportunities`. By just looking at the names, I can't be sure which of these is querying the db and which of the mare not.

File: auction-server/src/opportunity/entities/opportunity_svm.rs [hunk 4]
```diff
@@ -292,4 +295,28 @@ impl OpportunitySvm {
             OpportunitySvmProgram::Limo(_) => vec![],
         }
     }
+
+    pub fn get_permission_key(router: Pubkey, permission_account: Pubkey) -> PermissionKey {
+        PermissionKey::from([router.to_bytes(), permission_account.to_bytes()].concat())
+    }
+}
+
+impl From<OpportunitySvm> for api::OpportunityDelete {
```
Comment (m30m) [inline auction-server/src/opportunity/entities/opportunity_svm.rs:304]: These conversions are lossy, so I prefer a specific method over using the From trait.
https://doc.rust-lang.org/std/convert/trait.From.html#when-to-implement-from

File: auction-server/src/opportunity/entities/opportunity_svm.rs [hunk 5]
```diff
@@ -292,4 +295,28 @@ impl OpportunitySvm {
...
+        api::OpportunityDelete::Svm(api::OpportunityDeleteSvm::V1(api::OpportunityDeleteV1Svm {
+            chain_id:           val.chain_id.clone(),
+            permission_account: val.permission_account,
+            router:             val.router,
+            program:            val.program.into(),
+        }))
+    }
+}
+
+impl From<OpportunitySvmProgram> for api::ProgramSvm {
```
Comment (m30m) [inline auction-server/src/opportunity/entities/opportunity_svm.rs:315]: Same here:
These conversions are lossy, so I prefer a specific method over using the From trait.
https://doc.rust-lang.org/std/convert/trait.From.html#when-to-implement-from

**Review by m30m — APPROVED** [review]

File: auction-server/src/opportunity/entities/opportunity_svm.rs (hunk 5 above)
Comment (danimhr) [reply auction-server/src/opportunity/entities/opportunity_svm.rs:315]: I think this one makes more sense, but I removed the one for `OpportunityDelete`

**Pull request merged**